### DISCUSSION TITLE
Airflow for controlling a Robotic Machine workflow ?

### DISCUSSION THREAD

**Discussion by jishnair** [apache/airflow#29448] [Q&A] [answered]
Could someone please suggest if Airflow is the right tool for the following use case?
1. I need to control a Robotic machine with a complex workflow. The workflow contains 100 + Steps(tasks) and takes 3 days to complete. 
2. The machine needs to be controlled from a Local Windows computer which may not always have internet. So Airflow should be installed Locally on Windows. 
3. A separate Desktop/Browser based UI is needed to configure the workflow parameters. Also, the UI needs to display the machine state and errors in real-time. 

Another alternative I'm thinking of is a State machine-based backend and a JS front end, but since the workflow is complicated I think the Airflow DAGs could make the workflow much simpler.

**Comment by potiuk** [comment, marked as answer]
Having a past experience with Robotics (I've been a robotics engineer in https://nomagic.ai ) I'd recommend building a workflow via ROS - especially that it has a lot of Robotics-related modules, libraries classes etc. https://www.ros.org/ 

But If you insist and already have some ready-to use tasks / python modules that are independent from ROS. Then yes Airflow seems like a good candidate to do the job of orchestrating such workflows. Airlfow is not  designed to do it - it mostly works on data intervals and data processing (that's why I think ROS is better suited for your case). It's ok for airflow to control  ad-hoc (one-off) workflow like yours. 

One of the problem to solve is Linux vs. Windows. Airflow does not natively work on Windows (yet) but you can run it in docker container or in WSL2 on Windows. There is a bit of challenge to interface with Windows-specific drivers if you do it (in both cases) likely nothing unsormountable.

I think ROS should be your first candidate to be honest.

**Reply by jishnair** [reply]
@potiuk Thanks for your input. ROS may not be suited for my use case, since our machine already has a gRPC interface. I don't need any drivers or access to any low-level features. My only concern with airflow is running it natively in windows and updating the airflow status in real-time to the front end.

**Reply by potiuk** [reply]
Airflow does not run natively on Windows. you need WSL2 or Docker. If that's not a problem for you, there are no technical hurdles, but it's up to you to check if it's really suitable.
